### Network Connections
Another starting point and deeper dive for associated processes are the systems network connections the following commands will cover how to enumerate connection data structures in memory
#### connections

In [1]:
python ../volatility3/vol.py -h

The below script needs to be able to find the current output cell; this is an easy method to get it.

Volatility 3 Framework 1.2.1
usage: volatility [-h] [-c CONFIG] [--parallelism [{processes,threads,off}]]
                  [-e EXTEND] [-p PLUGIN_DIRS] [-s SYMBOL_DIRS] [-v] [-l LOG]
                  [-o OUTPUT_DIR] [-q] [-r RENDERER] [-f FILE]
                  [--write-config] [--clear-cache] [--cache-path CACHE_PATH]
                  [--offline] [--single-location SINGLE_LOCATION]
                  [--stackers [STACKERS [STACKERS ...]]]
                  [--single-swap-locations [SINGLE_SWAP_LOCATIONS [SINGLE_SWAP_LOCATIONS ...]]]
                  plugin ...

An open-source memory forensics framework

optional arguments:
  -h, --help            Show this help message and exit, for specific plugin
                        options use 'volatility <pluginname> --help'
  -c CONFIG, --config CONFIG
                        Load the configuration from a json file
  --parallelism [{processes,threads,off}]
                        Enables parallelism (defaults to off if no argument
       

    windows.envars.Envars
                        Display process environment variables
    windows.filescan.FileScan
                        Scans for file objects present in a particular windows
                        memory image.
    windows.getservicesids.GetServiceSIDs
                        Lists process token sids.
    windows.getsids.GetSIDs
                        Print the SIDs owning each process
    windows.handles.Handles
                        Lists process open handles.
    windows.info.Info   Show OS & kernel details of the memory sample being
                        analyzed.
    windows.malfind.Malfind
                        Lists process memory ranges that potentially contain
                        injected code.
    windows.memmap.Memmap
                        Prints the memory map
    windows.modscan.ModScan
                        Scans for modules present in a particular windows
                        memory image.
    windows.modules.Modules
            

In [12]:
python ../volatility3/vol.py -q -f E:\APT.img windows.netstat.NetStat

Volatility 3 Framework 1.2.1

Offset	Proto	LocalAddr	LocalPort	ForeignAddr	ForeignPort	State	PID	Owner	Created
Traceback (most recent call last):
  File "../volatility3/vol.py", line 10, in <module>
    volatility3.cli.main()
  File "C:\Users\analyst\Desktop\IncidentResponse\volatility3\volatility3\cli\__init__.py", line 625, in main
    CommandLine().run()
  File "C:\Users\analyst\Desktop\IncidentResponse\volatility3\volatility3\cli\__init__.py", line 333, in run
    renderers[args.renderer]().render(constructed.run())
  File "C:\Users\analyst\Desktop\IncidentResponse\volatility3\volatility3\cli\text_renderer.py", line 178, in render
    grid.populate(visitor, outfd)
  File "C:\Users\analyst\Desktop\IncidentResponse\volatility3\volatility3\framework\renderers\__init__.py", line 211, in populate
    for (level, item) in self._generator:
  File "C:\Users\analyst\Desktop\IncidentResponse\volatility3\volatility3\framework\plugins\windows\netstat.py", line 422, in _generator
    netscan_sy

In [13]:
python ../volatility3/vol.py -q -f E:\APT.img windows.netscan.NetScan

Volatility 3 Framework 1.2.1

Offset	Proto	LocalAddr	LocalPort	ForeignAddr	ForeignPort	State	PID	Owner	Created
Traceback (most recent call last):
  File "../volatility3/vol.py", line 10, in <module>
    volatility3.cli.main()
  File "C:\Users\analyst\Desktop\IncidentResponse\volatility3\volatility3\cli\__init__.py", line 625, in main
    CommandLine().run()
  File "C:\Users\analyst\Desktop\IncidentResponse\volatility3\volatility3\cli\__init__.py", line 333, in run
    renderers[args.renderer]().render(constructed.run())
  File "C:\Users\analyst\Desktop\IncidentResponse\volatility3\volatility3\cli\text_renderer.py", line 178, in render
    grid.populate(visitor, outfd)
  File "C:\Users\analyst\Desktop\IncidentResponse\volatility3\volatility3\framework\renderers\__init__.py", line 211, in populate
    for (level, item) in self._generator:
  File "C:\Users\analyst\Desktop\IncidentResponse\volatility3\volatility3\framework\plugins\windows\netscan.py", line 282, in _generator
    netscan_sy

In [11]:
volatility_2.6_win64_standalone.exe -h

Volatility Foundation Volatility Framework 2.6
Usage: Volatility - A memory forensics analysis platform.

Options:
  -h, --help            list all available options and their default values.
                        Default values may be set in the configuration file
                        (/etc/volatilityrc)
  --conf-file=.volatilityrc
                        User based configuration file
  -d, --debug           Debug volatility
  --plugins=PLUGINS     Additional plugin directories to use (semi-colon
                        separated)
  --info                Print information about all registered objects
  --cache-directory=C:\Users\analyst/.cache\volatility
                        Directory where cache files are stored
  --cache               Use caching
  --tz=TZ               Sets the (Olson) timezone for displaying timestamps
                        using pytz (if installed) or tzset
  -f FILENAME, --filename=FILENAME
                        Filename to use when opening an image


		volshell       	Shell in the memory image
		windows        	Print Desktop Windows (verbose details)
		wintree        	Print Z-Order Desktop Windows Tree
		wndscan        	Pool scanner for window stations
		yarascan       	Scan process or kernel memory with Yara signatures


In [7]:
volatility_2.6_win64_standalone.exe -f E:\APT.img imageinfo

Volatility Foundation Volatility Framework 2.6
INFO    : volatility.debug    : Determining profile based on KDBG search...
          Suggested Profile(s) : WinXPSP2x86, WinXPSP3x86 (Instantiated with WinXPSP2x86)
                     AS Layer1 : IA32PagedMemoryPae (Kernel AS)
                     AS Layer2 : FileAddressSpace (E:\APT.img)
                      PAE type : PAE
                           DTB : 0x319000L
                          KDBG : 0x80545b60L
          Number of Processors : 1
     Image Type (Service Pack) : 3
                KPCR for CPU 0 : 0xffdff000L
             KUSER_SHARED_DATA : 0xffdf0000L
           Image date and time : 2009-05-05 19:28:57 UTC+0000
     Image local date and time : 2009-05-05 15:28:57 -0400


In [6]:
volatility_2.6_win64_standalone.exe -f E:\APT.img --profile=WinXPSP2x86 connections

Volatility Foundation Volatility Framework 2.6
Offset(V)  Local Address             Remote Address            Pid
---------- ------------------------- ------------------------- ---
0x81e611f8 192.168.157.10:1053       218.85.133.23:89          796


Performaing a whois lookup on the IP `218.85.133.23`, from domain tools located here: http://whois.domaintools.com/218.85.133.23

```
IP Information for 218.85.133.23
 Quick Stats
IP Location	China China Fuzhou Chinanet Fujian Province Network
ASN	China AS133775 CHINATELECOM-FUJIAN-XIAMEN-IDC1 Xiamen, CN (registered Oct 28, 2014)
Whois Server	whois.apnic.net
IP Address	218.85.133.23
inetnum:        218.85.0.0 - 218.86.127.255
netname:        CHINANET-FJ
descr:          CHINANET Fujian province network
descr:          Data Communication Division
descr:          China Telecom
country:        CN
admin-c:        CH93-AP
tech-c:         CA67-AP
mnt-by:         MAINT-CHINANET
mnt-lower:      MAINT-CHINANET-FJ
status:         ALLOCATED NON-PORTABLE
last-modified:  2008-09-04T06:51:23Z
source:         APNIC

role:           CHINANETFJ IP ADMIN
address:        7,East Street,Fuzhou,Fujian,PRC
country:        CN
phone:          +86-591-83309761
fax-no:         +86-591-83371954
e-mail:         
remarks:        send spam reports  and abuse reports
remarks:        to 
remarks:        Please include detailed information and
remarks:        times in UTC
admin-c:        FH71-AP
tech-c:         FH71-AP
nic-hdl:        CA67-AP
remarks:        www.fjtelecom.com
notify:         
mnt-by:         MAINT-CHINANET-FJ
last-modified:  2011-12-06T00:10:50Z
source:         APNIC

person:         Chinanet Hostmaster
nic-hdl:        CH93-AP
e-mail:         
address:        No.31 ,jingrong street,beijing
address:        100032
phone:          +86-10-58501724
fax-no:         +86-10-58501724
country:        CN
mnt-by:         MAINT-CHINANET
last-modified:  2014-02-27T03:37:38Z
source:         APNIC
```

The connection was open to port 89 which is listed as SU MIT Telnet Gateway as the default service and because IE should be using 80 or 443 for HTTP and HTTPS connections. 

#### connscan
As with `pslist` and `psscan` volatility uses the `scan` naming convention to signify the plugin is performing a forensic search of the memory structure not just enumerating what the operating system would see. In this plugin the analysis is enumerating connection data structures from memory forensicly. 

In [8]:
volatility_2.6_win64_standalone.exe -f E:\APT.img --profile=WinXPSP2x86 connscan

Volatility Foundation Volatility Framework 2.6
Offset(P)  Local Address             Remote Address            Pid
---------- ------------------------- ------------------------- ---
0x0205ece0 192.168.157.10:1050       222.128.1.2:443           1672
0x020611f8 192.168.157.10:1053       218.85.133.23:89          796
0x032c01f8 192.168.157.10:1053       218.85.133.23:89          796
0x0337dce0 192.168.157.10:1050       222.128.1.2:443           1672
0x08a4ace0 192.168.157.10:1050       222.128.1.2:443           1672
0x18200ce0 192.168.157.10:1050       222.128.1.2:443           1672


This plugin exectuion finds another proess with an remote connection which was not previously found with the `connections` command. PID 1672 is explorer.exe which should not have any open connections to the Internet normally. Here is whois informaotin for the IP `222.128.1.2`:


```

IP Location	China China Beijing China Unicom Beijing Province Network
ASN	China AS4808 CHINA169-BJ China Unicom Beijing Province Network, CN (registered Jan 09, 1996)
Whois Server	whois.apnic.net
IP Address	222.128.1.2
inetnum:        222.128.0.0 - 222.131.255.255
netname:        UNICOM-BJ
descr:          China Unicom Beijing province network
descr:          China Unicom
country:        CN
admin-c:        CH1302-AP
tech-c:         SY21-AP
mnt-by:         APNIC-HM
mnt-lower:      MAINT-CNCGROUP-BJ
mnt-routes:     MAINT-CNCGROUP-RR
mnt-irt:        IRT-CU-CN
status:         ALLOCATED PORTABLE
last-modified:  2013-08-08T23:17:26Z
source:         APNIC

irt:            IRT-CU-CN
address:        No.21,Financial Street
address:        Beijing,100033
address:        P.R.China
e-mail:         
abuse-mailbox:  
admin-c:        CH1302-AP
tech-c:         CH1302-AP
auth:           # Filtered
mnt-by:         MAINT-CNCGROUP
last-modified:  2017-10-23T05:59:13Z
source:         APNIC

person:         ChinaUnicom Hostmaster
nic-hdl:        CH1302-AP
e-mail:         
address:        No.21,Jin-Rong Street
address:        Beijing,100033
address:        P.R.China
phone:          +86-10-66259764
fax-no:         +86-10-66259764
country:        CN
mnt-by:         MAINT-CNCGROUP
last-modified:  2017-08-17T06:13:16Z
source:         APNIC

person:         sun ying
address:        fu xing men nei da jie 97, Xicheng District
address:        Beijing 100800
country:        CN
phone:          +86-10-66030657
fax-no:         +86-10-66078815
e-mail:         
nic-hdl:        SY21-AP
mnt-by:         MAINT-CNCGROUP-BJ
last-modified:  2009-06-30T08:42:48Z
source:         APNIC

route:          222.128.0.0/14
descr:          China Unicom Beijing Province Network
country:        CN
origin:         AS4808
mnt-by:         MAINT-CNCGROUP-RR
last-modified:  2016-05-20T01:24:03Z
source:         APNIC
```

#### sockets
As with process enumeration there are several ways to enumerate network connections from memory. Below is another plugin, looks for and parses a singly-linked list of socket structures from the tcpip.sys module and only works on XP and 2003 systems.

In [9]:
volatility_2.6_win64_standalone.exe -f E:\APT.img --profile=WinXPSP2x86 sockets

Volatility Foundation Volatility Framework 2.6
Offset(V)       PID   Port  Proto Protocol        Address         Create Time
---------- -------- ------ ------ --------------- --------------- -----------
0x822ef510      796   1053      6 TCP             0.0.0.0         2009-05-05 19:29:06 UTC+0000
0x81fa9008     1212   1900     17 UDP             192.168.157.10  2009-04-16 16:10:20 UTC+0000
0x822d25b0        4      0     47 GRE             0.0.0.0         2009-05-05 19:28:35 UTC+0000
0x81e59748      716    500     17 UDP             0.0.0.0         2009-04-16 16:10:16 UTC+0000
0x822d7220        4    139      6 TCP             192.168.157.10  2009-04-16 16:10:00 UTC+0000
0x8206ce98      464   1032      6 TCP             127.0.0.1       2009-04-16 16:10:21 UTC+0000
0x82305008        4    445      6 TCP             0.0.0.0         2009-04-16 16:10:00 UTC+0000
0x81df5298      968    135      6 TCP             0.0.0.0         2009-04-16 16:10:07 UTC+0000
0x82305258        4    137     17 UDP

From this we can see PID 796 is listing on port 1053 at `2009-05-05 19:29:06 UTC+0000`. Giving you another data point for your analysis, especially for temporally based enumeration.  

In [10]:
volatility_2.6_win64_standalone.exe -f E:\APT.img --profile=WinXPSP2x86 sockscan

Volatility Foundation Volatility Framework 2.6
Offset(P)       PID   Port  Proto Protocol        Address         Create Time
---------- -------- ------ ------ --------------- --------------- -----------
0x01f354c8     1088    123     17 UDP             127.0.0.1       2009-04-16 16:10:20 UTC+0000
0x01ff5298      968    135      6 TCP             0.0.0.0         2009-04-16 16:10:07 UTC+0000
0x02059748      716    500     17 UDP             0.0.0.0         2009-04-16 16:10:16 UTC+0000
0x021a9008     1212   1900     17 UDP             192.168.157.10  2009-04-16 16:10:20 UTC+0000
0x021a9658     1088    123     17 UDP             192.168.157.10  2009-04-16 16:10:20 UTC+0000
0x0226ce98      464   1032      6 TCP             127.0.0.1       2009-04-16 16:10:21 UTC+0000
0x023514f8     1212   1900     17 UDP             127.0.0.1       2009-04-16 16:10:20 UTC+0000
0x023554c8     1088   1028     17 UDP             127.0.0.1       2009-04-16 16:10:20 UTC+0000
0x02365320      716      0    255 Res